### PR TITLE
add support for dynamic values

### LINKED ISSUE #14: support dynamic values
Hi!

I authored and maintain a library called [`react-shed`](https://github.com/tedconf/react-shed/tree/emotion), which uses a css-in-js library to allow users to define styles that are then applied to elements like this:

```jsx
<Shed
  component="div"
  mx="5"
>output is margin-left: sizes(5), margin-right: sizes(5)</Shed>
```

is is possible to calculate values based on props with _picostyle_?

Thanks!

### PR DESCRIPTION
**Pull request by VinSpee**
pass `data` to `decls` which allows for the calculation of rules based on component properties

resolves #14

### CODE REVIEW

**Comment by VinSpee** [comment]
codepen demo: https://codepen.io/VinSpee/pen/qPbdJp?editors=0011

**Comment by maxholman** [comment]
Unfortunately this change would revert the benefits of the `decls` and class name reuse implemented in #13

**Comment by VinSpee** [comment]
@maxholman oohhhhhhhhh

what if i check if decls is a function and act accordingly?

**Comment by maxholman** [comment]
Looks like it could work. You might be able to squeeze a few more bytes out by caching the result of the `typeof` check.

I think you can also do this sort of thing without any code changes to picostyle:

```javascript
const DynText = (props, children) => ps("h1")({
  fontSize: "64px",
  lineHeight: "1.5",
  cursor: "pointer",
  color: props.c,
})(props, children);
```

**Comment by VinSpee** [comment]
oh wow yeah I can do it without changes to core.

I'll make your suggested byte-squeezing check and leave this for the maintainers to close or merge depending on their preferences. I don't think it'd be _bad_ to have styled-component-like API, but I understand this workaround.

File: src/index.js [hunk 1]
```diff
@@ -38,9 +38,17 @@ function parse (decls, child, media, className) {
...
+      var declsType = typeof decls;
+      if (declsType !== 'function') {
+        parsed = parse(decls)
+      }
       return function (data, children) {
+        if (declsType === 'function') {
+          parsed = parse(decls(data))
+        }
         data = data || {}
+        var parsed = parse(decls(data))
```
Comment (morishitter) [inline src/index.js]: `var` is not necessary in this line.

File: src/index.js [hunk 2]
```diff
@@ -38,9 +38,17 @@ function parse (decls, child, media, className) {
 export default function (h) {
   return function (tag) {
     return function (decls) {
-      var parsed = parse(decls)
+      var parsed;
+      var declsType = typeof decls;
```
Comment (morishitter) [inline src/index.js]: Please remove `;` on L41 and L42.

**Comment by JorgeBucaran** [comment]
@VinSpee @maxholman @morishitter I am confused now. I thought these changes were not needed? 🤔

**Comment by VinSpee** [comment]
only needed if we want that API (which i think is nice to have)

**Comment by morishitter** [comment]
Sorry my bad. Checking to the type of `decls` may be nonsense 🤔  https://github.com/picostyle/picostyle/pull/15/files#diff-1fdf421c05c1140f6d71444ea2b27638R51

**Comment by VinSpee** [comment]
what do you mean "nonsense"?

**Comment by JorgeBucaran** [comment]
@VinSpee Maybe he meant unneeded. 

@morishitter If you want you can revert this or implement it again yourself? To be honest, if this can be done outside core I would prefer it that way.

**Comment by morishitter** [comment]
> Maybe he meant unneeded.

Yes! 

@maxholman I'm not good at English. Please be generous in finding my rude expressions.

@JorgeBucaran I will implement it again without to revert :)

**Comment by VinSpee** [comment]
@morishitter I understood that english was not your native language - no offense taken at all, I was just looking for clarification

**Comment by morishitter** [comment]
@VinSpee I meant, [this line](https://github.com/picostyle/picostyle/pull/15/files#diff-1fdf421c05c1140f6d71444ea2b27638R51) is run regardless of the type of `decls` so this lines occurs an error if `decls` is not `function`.

**Pull request merged**